runtime/nvqir/custatevec/CuStateVecCircuitSimulator.cu

```diff
@@ -369,10 +368,93 @@ public:
     }
   }
 
+  /// @brief Compute the operator expectation value, with respect to 
+  /// the current state vector, directly on GPU with the 
+  /// given the operator matrix and target qubit indices. 
+  auto getExpectationFromOperatorMatrix(const std::complex<double> *matrix,
+                                        const std::vector<std::size_t> &tgts) {
+    void *extraWorkspace = nullptr;
+    size_t extraWorkspaceSizeInBytes = 0;
+
+    // Convert the size_t tgts into ints
+    std::vector<int> tgtsInt(tgts.size());
+    std::transform(tgts.begin(), tgts.end(), tgtsInt.begin(),
+                   [&](std::size_t x) { return static_cast<int>(x); });
+    // our bit ordering is reversed.
+    std::reverse(tgtsInt.begin(), tgtsInt.end());
+    size_t nIndexBits = nQubitsAllocated;
+
+    // check the size of external workspace
+    HANDLE_ERROR(custatevecComputeExpectationGetWorkspaceSize(
+        handle, cuStateVecCudaDataType, nIndexBits, matrix,
+        cuStateVecCudaDataType, CUSTATEVEC_MATRIX_LAYOUT_ROW, tgts.size(),
+        cuStateVecComputeType, &extraWorkspaceSizeInBytes));
+
+    if (extraWorkspaceSizeInBytes > 0) {
+      HANDLE_CUDA_ERROR(cudaMalloc(&extraWorkspace, extraWorkspaceSizeInBytes));
+    }
+
+    double expect;
+
+    // compute expectation
+    HANDLE_ERROR(custatevecComputeExpectation(
+        handle, deviceStateVector, cuStateVecCudaDataType, nIndexBits, &expect,
+        CUDA_R_64F, nullptr, matrix, cuStateVecCudaDataType,
+        CUSTATEVEC_MATRIX_LAYOUT_ROW, tgtsInt.data(), tgts.size(),
+        cuStateVecComputeType, extraWorkspace, extraWorkspaceSizeInBytes));
+    if (extraWorkspaceSizeInBytes)
+      HANDLE_CUDA_ERROR(cudaFree(extraWorkspace));
+
+    return expect;
+  }
+
+  /// @brief We can compute Observe from the matrix for a
+  /// reasonable number of qubits, otherwise we should compute it
+  /// via sampling
+  bool canHandleObserve() override {
+    // Do not compute <H> from matrix if shots based sampling requested
+    if (executionContext &&
+        executionContext->shots != static_cast<std::size_t>(-1)) {
+      return false;
+    }
+
+    /// Seems that FP32 is faster with 
+    /// custatevecComputeExpectationsOnPauliBasis
+    if constexpr (std::is_same_v<ScalarType, float>) {
+      return false;
+    }
+
+    // FIXME Study the tradeoff for this as NQubits gets larger...
+    // Maybe the 14 qubit range
+    if (nQubitsAllocated > 14) {
+      return false;
+    }
```

Comment:
Instead of using a magic number, create a `constexpr`, say `max_observe_qubits` or something.

Out of curiosity why 14? Does this depends on the user's system? Might just as well, live it unbounded and let the system complain or the user get bored of waiting. We should provide a way for the user to configure this number.

Reply:
I agree that this needs a bit more thought. The question here is under what circumstances should we compute <psi | H | psi> directly from the state vector and matrix data for H, or if we should compute it from just the state amplitudes and parity of the measured bits. We want to choose correctly, and it depends on the size of the state vector and the size of the matrix H. For the benchmarks above, the matrix is small, so <psi | H | psi> is faster up to about 14 qubits.

Reply:
I think that for now, I should remove this check and instead add an environment variable check that allows one to switch between the 2 ways of expectation value calculation.